source/extensions/http/header_validators/envoy_default/http1_header_validator.cc

```diff
@@ -0,0 +1,420 @@
+  // such a request MUST generate a new Host field-value based on the received
+  // request-target rather than forward the received Host field-value.
+  // ...
+  // If the target URI includes an authority component, then a client MUST send a
+  // field-value for Host that is identical to that authority component,
+  // excluding any userinfo subcomponent and its "@" delimiter (Section 2.7.1).
+  //
+  // TODO(meilya) - This needs to be implemented after we have path normalization so that we can
+  // parse the :path form and compare the authority component of the path against the :authority
+  // header.
+  auto is_connect_method = header_map.method() == header_values_.MethodValues.Connect;
+  auto is_options_method = header_map.method() == header_values_.MethodValues.Options;
+
+  if (!is_connect_method && path.empty()) {
+    // The :path is required for non-CONNECT requests.
+    return {RequestHeaderMapValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidUrl};
+  }
+
+  auto path_is_asterisk = path == "*";
+  auto path_is_absolute = path.empty() ? false : path.at(0) == '/';
+
+  // HTTP/1.1 allows for a path of "*" when for OPTIONS requests, based on RFC
+  // 9112, https://www.rfc-editor.org/rfc/rfc9112.html#section-3.2.4:
+  //
+  // The asterisk-form of request-target is only used for a server-wide OPTIONS
+  // request
+  // ...
+  // asterisk-form  = "*"
+  if (!is_options_method && path_is_asterisk) {
+    return {RequestHeaderMapValidationResult::Action::Reject,
+            UhvResponseCodeDetail::get().InvalidUrl};
+  }
+
+  // Step 2: Validate Transfer-Encoding and Content-Length headers.
+  // HTTP/1.1 disallows a Transfer-Encoding and Content-Length headers,
+  // https://www.rfc-editor.org/rfc/rfc9112.html#section-6.2:
+  //
+  // A sender MUST NOT send a Content-Length header field in any message that
+  // contains a Transfer-Encoding header field.
+  //
+  // The http1_protocol_options.allow_chunked_length config setting can
+  // override the RFC compliance to allow a Transfer-Encoding of "chunked" with
+  // a Content-Length set. In this exception case, we remove the Content-Length
+  // header.
+  if (header_map.TransferEncoding()) {
+    // CONNECT methods must not contain any content so reject the request if Transfer-Encoding or
+    // Content-Length is provided, per RFC 9110,
+    // https://www.rfc-editor.org/rfc/rfc9110.html#section-9.3.6:
+    //
+    // A CONNECT request message does not have content. The interpretation of data sent after the
+    // header section of the CONNECT request message is specific to the version of HTTP in use.
+    if (is_connect_method) {
+      return {RequestHeaderMapValidationResult::Action::Reject,
+              Http1ResponseCodeDetail::get().TransferEncodingNotAllowed};
+    }
+
+    if (header_map.ContentLength() &&
+        header_map.TransferEncoding()->value() == header_values_.TransferEncodingValues.Chunked) {
```

Comment:
I think the original code with `hasChunkedTransferEncoding()` was correct here. Otherwise this is a change in behavior.
For example previously request with:
```
transfer-encoding: identity,chunked
content-length: 11
```
Would fail with 400 and content_length_and_chunked_not_allowed details, now it will fail in the transfer-encoding check and (after it is fixed) produce 501 for unsupported encoding. I think it is worthwhile to preserve the original behavior and keep the check that supports multiple tokens in the transfer-encoding.